src/core_plugins/kibana/public/dashboard/actions/panels.js

```diff
@@ -0,0 +1,51 @@
+import { createAction } from 'redux-actions';
+
+export const deletePanel = createAction('DELETE_PANEL', panelId => panelId);
+
+export const embeddableRenderFinished =
+  createAction('EMBEDDABLE_RENDER_FINISHED', (panelId, embeddable) => ({ embeddable, panelId }));
+
+export const embeddableRenderError =
+  createAction('EMBEDDABLE_RENDER_ERROR', (panelId, error) => ({ panelId, error }));
+
+export const updatePanel = createAction('UPDATE_PANEL', panel => panel);
+
+/**
+ * @param panels {Array<PanelState>}
+ * @return {Object}
+ */
+export const updatePanels = createAction('UPDATE_PANELS', panels => {
+  const panelsMap = {};
+  panels.forEach(panel => {
+    panelsMap[panel.panelIndex] = panel;
+  });
+  return panelsMap;
+});
+
+/**
+ *
+ * @param embeddableHandler {EmbeddableHandler}
+ * @param panelElement {Node}
+ * @param panelId {string}
+ * @param containerApi {ContainerAPI}
+ * @return {function(*, *)}
+ */
+export function renderEmbeddable(embeddableHandler, panelElement, panelId, containerApi) {
+  return (dispatch, getState) => {
+    const { dashboardState } = getState();
```

Comment:
I definitely understand the sentiment, but I ran into this same pattern and the reason I wanted to stick with `getState` in the action is because I didn't want state changes for that particular state to cause a re-render of my component unnecessarily.  So in this case, [it doesn't seem](https://github.com/Stacey-Gammon/kibana/blob/461c4ede8c4b8175fa4e93fe89886214036660ea/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel_container.js#L11) like the component needs to know about that particular part of the state tree so adding it here just to make it available in the presentational component so it can call the action with it seems not great to me. 

Any thoughts on that?

Reply:
Why is a re-render a problem? If this state updates multiple times per second, I might agree (though, I'd likely push towards implementing a `shouldComponentUpdate` if performance on this component is that important and we're hitting rendering slowness). I still prefer injecting it, as I feel it makes the code clearer.

Reply:
Nothing inherently wrong, but just felt off to do it that way. I definitely understand the hesitation around calling `getState` within the action, but I'm not (yet) convinced it's an anti pattern, but always willing to hear counter arguments.

Reply:
I'm inclined to agree with @chrisronline on this one, though perhaps because it makes my life easier.  If the only reason your putting a prop in a component is to pass it to a function, which goes back to the container, seems like an unnecessary loop.

I was actually surprised I couldn't get the state inside of `mapDispatchToProps` - that would solve this.  I have the whole state tree inside of `mapStateToProps`, why not also have it accessible in `mapDispatchToProps`?  Unless it is and I don't know about it?  Because if it was I could do something like:

```

const mapDispatchToProps = (dispatch, { embeddableHandler, panelId, getContainerApi }, { dashboardState }) => ({
  renderEmbeddable: (panelElement) => {
    const panel = getPanel(dashboardState, panelId);
    return renderEmbeddable(embeddableHandler, panelElement, panel, getContainerApi());
  }
}
```

But because I couldn't do that, I just pushed the `getState` inside the action.

Seems silly to make the DashboardPanelContainer require panel as a prop instead of just panelId.  But, that's just my gut feel.

Reply:
You can try messing with the `mergeProps` discussed [here](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options) but from what I remember when playing with it, it didn't stop the unnecessary rerender, but feel free to try it out